### DISCUSSION TITLE
Proof assistant in Symja

### DISCUSSION THREAD

**Discussion by TiagoCavalcante** [axkr/symja_android_library#454] [Ideas]
The [few] proof assistants available in the "market" (all I know are free, and probably all that exists are free, as they won't have [m]any buyers).

Proof assistants aren't used widely mostly because their syntax is weird, and they have a steep learning curve.

WL (and Symja, and Mathics) have a syntax that is easier to use on the other hand. After searching I have found that [almost?] nobody uses WL for proof assistant, probably because it doesn't support it. And so I had the idea: can a proof assistant be developed in Symja?

The core of a proof assistant is to follow logical steps, and check if their logic is correct based on a set of axioms (axioms like ∀x,y . ∀z (z ∈ x ∧ z ∈ y) ⟺ x = y or 1+1=2).

I think that a library could provide functions that receive assumptions and logical expressions, and check if every step makes sense.

An idea of how it would be:
```wl
(* there would be 2 types of proof: DirectionalProof and ProofByContradiction *)
ProofByContradiction[ (* proof that sqrt(prime) is irrational *)
  { (* assumptions *)
    PrimeQ[p],
    IntegerQ[m],  (* an assumption of an undefined symbol "creates" that symbol *)
                          (* that can be done with the pattern matching system *)
                          (* basic types like Integer, Real, and Complex would have those rules *)
    IntegerQ[n],  (* the symbols defined above can be used in the assumptions below them *)
    GCD[m, n] == 1, (* assumptions are boolean values *)
    m/n == Sqrt[p]
  },
  { (* steps *)
    Step[
      m^2/n^2 == Sqrt[p]^2, (* each step contains 1 manipulation of any assumtion *)
      m^2/n^2 == p, (* basic equation manipulation would need to be implemented *)
      m^2 == p * n^2,
      Divisible[m^2, p], (* not so basic manipulation *)
      Divisible[m, p],
      Lemma[m == p * k] (* create a new "assumption" from the last equation *)
    ],
    Step[
      m^2/n^2 == Sqrt[p]^2,
      m^2/n^2 == p,
      m^2 == p * n^2,
      (p * k)^2 == p * n^2, (* here it becomes clear why each step need to contain only 1 manipulation *)
                                       (* so we can easily replace variables with rules defined by the assumptions and lemmas *)
      p^2 * k^2 = p * n^2,
      p * k^2 = n,
      Divisible[n, p * k^2],
      Divisible[n, p] (* on each step the assumptions are checked *)
      (* this is the last step of the proof because here the library finds a contradiction in an assumption, so proof by contradictio ∎ *)
    ]
  }
]
```

This is just an idea and probably my "design" can be improved by 100%.

If a good "design" is created, I'd like to help to implement it.

**Comment by axkr** [comment]
A first starting point can be to integrate existing Satisfiability modulo theories:
- https://en.wikipedia.org/wiki/Satisfiability_modulo_theories

An umbrella project for the different solvers seems to be "JavaSMT - Unified Java API for SMT solvers.":
- https://github.com/sosy-lab/java-smt

For boolean logic the "LogicNG" project is already used in Symja as a dependency:
- https://github.com/logic-ng/LogicNG

See `BooleanFunctions.java`
- https://github.com/axkr/symja_android_library/blob/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/builtin/BooleanFunctions.java

**Reply by raner** [reply]
Is there a way to leverage the existing LogicNG dependency to create something like a simple theorem prover?
For example, let's say function `a` expects a natural number that is not zero. Further, function `b` produces a natural number that is guaranteed to be greater than 1. Passing function `b`'s result as input to function `a` should be valid, because `n > 1` is a stronger assertion that implies `n != 0` (for natural numbers). Similary, a function might expect statement `x` to be true for its input. Another function might produce a result about which can be said `x || y`, but from the surrounding context we might also know that `!y` is true. Again, passing the value should be valid, but takes application of some logic rules to prove. Does Symja have support for logic proofs like that?

**Reply by axkr** [reply]
@czengler is this possible?
> Is there a way to leverage the existing LogicNG dependency to create something like a simple theorem prover?
